### ISSUE TITLE
[Feature-request] hillshade raster layer

### ISSUE THREAD

**Issue by farfromrefug** [CartoDB/mobile-sdk#356]
Thanks to the work of Mapbox we can have a layer which draws dynamic hillshade (with orientation, lightning ...) based on raster tile.

Mabpox implemented it [here](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/layers/render_hillshade_layer.cpp) with a shader.

I think it can be translated to Carto. I started to look at it but quite lost right now. I see that you use shader for 3D elements for example. Have to see how to translate that yet.

Another quite interesting thing with this is that you can actually get the elevation from that!
So it means:
* elevation on click
* possibility to compute elevation profile without using Valhalla.

If you are interested here is an example mbtiles of such rgb tiles (for france):
https://www.data.gouv.fr/fr/datasets/r/0baf6d7c-dc91-4682-b04f-5e0f5413b975

**Comment by mtehver** [comment]
Tagging this with 'idea-maybe-future'.

**Comment by farfromrefug** [comment]
As a not here i already managed to compute elevation profile based on reading the mbtiles.
I put the javascript code i use here as a reference.
Now only the shader drawing remain (the toughest part obviously)
```javascript
const zoom_level = 11;
        const tilesIndexed = {};

        // sort the points per tile
        _points.forEach((p, index) => {
            const tile = latLngToTileXY(p.lat, p.lon, zoom_level, 512);
            const key = tile.x + '' + tile.y;
            if (!tilesIndexed[key]) {
                tilesIndexed[key] = { x: tile.x, y: tile.y, poses: [] };
            }
            tilesIndexed[key].poses.push({ index, pixelX: tile.pixelX, pixelY: tile.pixelY, ...p });
        });
        const result = [];
        // request all the necessary tiles from db
        await Promise.all(
            Object.keys(tilesIndexed).map(k => {
                const t = tilesIndexed[k];
                return db
                    .get(
                        `SELECT tile_data FROM tiles WHERE zoom_level=${zoom_level} AND tile_row=${(1 << zoom_level) -
                            1 -
                            t.y} and tile_column=${t.x}`
                    )
                    .then(row => {
                        if (row && row.tile_data) {
                            const data = row.tile_data as any;

                            // decode into android bitmap to get access to the pixels
                            const bmp = android.graphics.BitmapFactory.decodeByteArray(data, 0, data.length);
                            t.poses.forEach(p => {
                                const pixel = bmp.getPixel(p.pixelX, p.pixelY);
                                const R = (pixel >> 16) & 0xff;
                                const G = (pixel >> 8) & 0xff;
                                const B = pixel & 0xff;
                                result[p.index] = {
                                    lat: p.lat,
                                    lon: p.lon,
                                    height: Math.round(-10000 + (R * 256 * 256 + G * 256 + B) * 0.1)
                                };
                            });
                            bmp.recycle();
                        }
                    });
            })
        );
        // transform the result into a profile we can use
        return this.computeProfileFromHeights(result);
```

**Comment by farfromrefug** [comment]
@mtehver i am looking at implementing this.
I need a bit of advice from you. This is what i intend to do:
* create a new `HillshadeDataSource`. No real question there
* create a new `HilllshadeRasterTileLayer` extending `RasterTileLayer` using a new render `HillshadeTileRenderer`
* create the new `HillshadeTileRenderer` extending `TileRenderer` to include and use the new shaders like the one from Mabpox.

Do you think that is the right way. Though the thing where i am lost is in the changes you recently made to the renderers. I dont see where i can "create" that render for that new tile layer.
Can you guide me a bit?

**Comment by farfromrefug** [comment]
Ok so found out a bit more.
It seems that the bitmap rendering is handled in [GTileRender](https://github.com/cartodb/mobile-carto-libs/blob/develop/vt/src/vt/GLTileRenderer.cpp#L1351)
So i guess i need to create a new method there `renderTileHillshadeBitmap` which uses a new shader like the one from [mapbox](https://github.com/mapbox/mapbox-gl-native/blob/901a96048755587fdd5d1332695d2021fc169591/src/mbgl/programs/gl/hillshade.cpp)

Then i would change `TileLayer` to add `getHillshadeBitmaps`
And finally implement my `CreateVectorTile` method to build a `TileLayer`

Does that seem good to you?

**Comment by mtehver** [comment]
Correct, the actual rendering is handled by `GLTileRenderer` in mobile-carto-libs repo. There are a few complications, like handling 'globe mode' correctly.

I have not really looked closely into this, but I think all this can be handled by a 'lighting shader' that  simply renderers alpha bitmap on top of existing layers. An example of lighting shader that is used for both raster and vector layers can be found here: https://github.com/CartoDB/mobile-sdk/blob/master/all/native/renderers/TileRenderer.cpp#L305

Assuming that is the case, then perhaps mobile-carto-libs does not have to be modified at all, changes are only required in `TileRenderer` and `RasterLayer` classes.

**Comment by farfromrefug** [comment]
@mtehver i managed to implement it.
I will create a PR, however there are some changes necessary in the carto lib.
The reason is double:
* my shader need to have access to `uBitmap`. But because of the concatenation order [here](https://github.com/CartoDB/mobile-carto-libs/blob/6cbb6b86c4efc00fd3681d7bc770cfead469e7dd/vt/src/vt/GLTileRenderer.cpp#L1873) that variable is not available. But you need that order for `applyLightning` to be defined before the light shader. Now there are 2 solutions for this:
   - split [this](https://github.com/CartoDB/mobile-carto-libs/blob/6cbb6b86c4efc00fd3681d7bc770cfead469e7dd/vt/src/vt/GLTileRendererShaders.h#L178) in 2 strings to be able to concatenate variables before the lighting shader.
   - pass `uBitmap` as a variable of `applyLighting` which i did for now.
* i need a shader variable depending on the tile id. However [here](https://github.com/CartoDB/mobile-sdk/blob/master/all/native/renderers/TileRenderer.cpp#L282) i can't seem to have access to the current drawn tile. Thus there i also had to pass it [here](https://github.com/CartoDB/mobile-carto-libs/blob/6cbb6b86c4efc00fd3681d7bc770cfead469e7dd/vt/src/vt/GLTileRenderer.cpp#L1361) to the shader and pass it as a arg of `applyLighting`

Would love to have your opinion before creating any PR. If there is a way to not modify the carto lib it would be better.

**Comment by mtehver** [comment]
Closing this, this is now implemented in 4.3.1 release.

A few comments:

* Main light direction affects the shading. The default light direction is from **below** and does not work very well. I suggest using `MapVec(0, 1, 1)` as a starting point (this is a light source from north at 45deg azimuth).

* The MapBox data source for height info uses upsampling starting from zoom 12. This results in a blocky artifacts at higher zoom levels. This is not something that the SDK can fix.
